### PR TITLE
add an elixir secret pull

### PR DESCRIPTION
**Pull request by craig-day**
Add an elixir library to pull secrets from the secrets sidecar. 

/cc @grosser @jonmoter @jcheatham 

@jcheatham can you give this a nice look since you're pretty familiar with elixir?

### CODE REVIEW

File: elixir/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# SamsonSecretPuller
+
+Gem to read secrets generated by the puller.
```
Comment (grosser) [inline elixir/README.md]: not `Gem` ;)

File: elixir/lib/samson_secret_puller.ex [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+      iex> SamsonSecretPuller.fetch_secrets!
+      [MYSQL_USER: "admin", MYSQL_PASS: "password"]
+
+  """
+  @spec fetch_secrets! :: Keyword.t
+  def fetch_secrets! do
+    case wait_for_secrets_to_appear do
+      {:ok, _}      -> read_secrets
+      {:error, msg} -> []
```
Comment (grosser) [inline elixir/lib/samson_secret_puller.ex]: ignore errors ... seems weird ...

File: elixir/lib/samson_secret_puller.ex [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    end
+  end
+
```
Comment (craig-day) [inline elixir/lib/samson_secret_puller.ex]: everything defined with `defp` is private. So there is only one public method, `fetch_secrets!` and the rest are private.

**Comment by grosser** [comment]
ahh sneaky :)

**Comment by craig-day** [comment]
anyone have any other thoughts? I pushed another commit just with some package information.  You can see the published package here: https://hex.pm/packages/samson_secret_puller

File: .gitignore [hunk 4]
```diff
@@ -1,6 +1,9 @@
 *.rbc
+mix.lock
```
Comment (grosser) [inline .gitignore]: ... won't that make it fail whenever a new version comes out and then whoever wanted to modify something unrelated has to figure out a elixir issue ?

File: elixir/LICENSE [hunk 5]
```diff
@@ -0,0 +1,18 @@
+Permission is hereby granted, free of charge, to any person obtaining
```
Comment (craig-day) [inline elixir/LICENSE]: that makes the release manager unhappy.

Comment (grosser) [inline elixir/LICENSE]: is that some elixir thing ?

Comment (grosser) [inline elixir/LICENSE]: ... then at least have some "MIT License" at the top so people know what this is

File: elixir/test/samson_secret_puller_test.exs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    File.rm_rf!   "./test/secrets" # Make sure each test starts with a clean slate
+    File.mkdir_p! "./test/secrets"
+    File.write!   "./test/secrets/FOO", "bar"
+    {:ok, []}
+  end
+
+  test "reads secrets from the given folder" do
+    File.touch!   "./test/secrets/.done"
+
+    assert SamsonSecretPuller.fetch_secrets!("./test/secrets") == [FOO: "bar"]
```
Comment (grosser) [inline elixir/test/samson_secret_puller_test.exs]: no assert_equal ?

**Comment by grosser** [comment]
👍

File: elixir/test/samson_secret_puller_test.exs (hunk 6 above)
Comment (craig-day) [inline elixir/test/samson_secret_puller_test.exs]: Nope, but the test framework fails in a much nicer way. It doesn't just say "assertion failed, no message given"

```
  1) test reads secrets from the given folder (SamsonSecretPullerTest)
     test/samson_secret_puller_test.exs:15
     Assertion with == failed
     code: [1, 2, 3] == [2, 3, 4]
     lhs:  [1, 2, 3]
     rhs:  [2, 3, 4]
     stacktrace:
       test/samson_secret_puller_test.exs:18
```

**Comment by grosser** [comment]
nice magic trick oO

**Comment by craig-day** [comment]
I don't know what magic you're talking about, but too late. Already merged. :D

**Comment by grosser** [comment]
the magic that the assert knows what both sides of the equal were ... so I
guess it gets compiled down into something like assert_equal

**Pull request merged**